[local_auth] Allow device authentication (pin/pattern/passcode)

kennethj opened the pull request:
## Description

Added `LocalAuthentication().authenticate()` method that allows users to authenticate using biometric (fingerprint, face) or device authentication (pin, pattern, passcode).

## Screenshots
![ios-auth](https://user-images.githubusercontent.com/862054/73473432-6212cd00-4352-11ea-96e2-73688da399af.gif)
<img width="476" alt="Screen Shot 2020-01-31 at 8 29 51 AM" src="https://user-images.githubusercontent.com/862054/73547057-db1e2d00-4403-11ea-8f02-fea02d17d6b2.png">
## Related Issues

This resolves the feature request on https://github.com/flutter/flutter/issues/49703

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [ ] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

jitenders859 commented:
on Android Can see the use password button but when clicking on it it closes the dialog and shows this message in console

FingerprintDialogFrag( 4198): Failed to check device credential. Parent handler not found.

kennethj commented:
@jitenders859 Is that on a physical device or with on an emulator?
I was having issues running this on android emulators, but it appears to work on physical devices. 
After a little research, I found there may be `KeyguardManager#isDeviceSecure()` check that could be added to avoid the error.

jitenders859 commented:
@kennethj I am using real android device and 
also can you show how to implement KeyguardManager#isDeviceSecure()

also in file

https://android.googlesource.com/platform/frameworks/support/+/refs/heads/androidx-recyclerview-release/biometric/src/main/java/androidx/biometric/Utils.java
From the source code we can see at line 82 that if the activity type is not of DeviceCredentialHandlerActivity it will fail

jitenders859 commented:
@kennethj 

![ezgif-3-b71f99e54d14](https://user-images.githubusercontent.com/41188075/75974844-79b71580-5efd-11ea-9290-7daa0e0ecb1d.gif)

In the above gif when pressed on authenticated the popup opens two times and when pressed on use password it closes on reopening can't see fingerprint icon and on authenticating with fingerprint didn't authenticate. after that opening second time it authenticates using fingerprint.

also is there any problem with sdk version below 28

kennethj commented:
@jitenders859 
This does appear to be an issue with API 28 and below. This feature only seems to work as intended on API 29 (Android 10). I am going to try to find time today to look into it more.

apgapg commented:
@kennethj In Android 6.0 with non-fingerprint device it doesnt allow to change verification method.
Device: **Samsung J2 Pro, SM-J210F android 6.0.1  without fingerprint**

Here's my implementation:
```
  local_auth:
    git:
      url: https://github.com/kennethj/flutter-plugins
      ref: feature/local_auth
      path: packages/local_auth

```
my expectation is to have pattern verification also which works fine in higher android version. Only in 6.0 it fails.

![IMG-20200603-WA0005](https://user-images.githubusercontent.com/13887407/83767584-74539d80-a69b-11ea-87e6-bcbb17270fc3.jpeg)

![Screenshot_2020-06-03-15-45-54-822_com workozy android demo](https://user-images.githubusercontent.com/13887407/83767536-63a32780-a69b-11ea-8c00-5229452ea0b8.jpg)

apgapg commented:
Any progress in this PR?
We really need pattern lock verification because some devices dont have fingerprint.

SergejSachs commented:
yea thats true, device password auth would be so good as a fallback

kdhanji87 commented:
Any progress in this PR?
We really need pattern lock verification because some devices dont have fingerprint would be a useful fallback

duncwillz commented:
@kennethj LGTM. @softmarshmallow @mehmetf Could you please help get this PR merged?. It wouldn't be worth reinventing the wheel locally. Thank you.

Zelosys commented:
Hello
Any ETA on this PR?

It's a necessary feature for our App, and we don't want to reinvent the Wheel and do it manual, when the plugin can do this soon by its own :)

SergejSachs commented:
Any update on this?

Zelosys commented:
It looks like after three months it's time to annoy the Google staff again.😀
Any updates on this? @mehmetf 
Is there anything we can do to speed this up and get it in?

mehmetf commented:
Is this PR actively maintained? It needs to be rebased. Please ping me and I will do a pass after someone claims ownership.

@Zelosys @kennethj

kennethj commented:
@mehmetf I will probably have some time in the next two weeks or so.

kennethj commented:
@mehmetf I think this is ready for review.

mehmetf commented on packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java:
```diff
@@ -75,24 +73,38 @@
+              .setTitle((String) call.argument("signInTitle"))
+              .setSubtitle((String) call.argument("biometricHint"))
+              .setConfirmationRequired((Boolean) call.argument("sensitiveTransaction"))
+              .setDeviceCredentialAllowed(true)
```
You can pass  `allowCredentials` to  `setDeviceCredentialAllowed` call and get rid  of the conditional.

kennethj replied:
There was an issue with setting both `setNegativeButtonText` and `setDeviceCredentialAllowed`. 


From android docs:
> Note that enabling this option replaces the negative button on the prompt with one that allows the user to authenticate with their device credential, making it an error to call setNegativeButton(java.lang.CharSequence, java.util.concurrent.Executor, android.content.DialogInterface.OnClickListener).

I will refactor to try to make it more obvious what is happening.

kennethj replied:
updated

mehmetf commented on packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java:
```diff
@@ -125,21 +137,27 @@ private void stop() {
-        break;
+        if (call.argument("useErrorDialogs")) {
+          showGoToSettingsDialog(
+              "Device credentials required",
```
These  strings should come from Dart just like the rest of the status strings used below.

kennethj replied:
Added to `AndroidAuthMessages` class

mehmetf commented on packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java:
```diff
@@ -213,19 +231,23 @@ public void onResume(@NonNull LifecycleOwner owner) {
     onActivityResumed(null);
   }
 
+  // hack: dialog opens twice on Android 8
```
Do we know why? Do the handlers below fire twice for some reason?

kennethj replied:
I am not sure. This happened a while ago and don't remember the specifics and I have not been able to reproduce the issue in emulators today.

mehmetf replied:
Let's remove it if it cannot be reproduced.

kennethj replied:
removed

mehmetf commented on packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java:
```diff
@@ -65,138 +85,275 @@ private LocalAuthPlugin(Activity activity) {
   @Override
   public void onDetachedFromEngine(FlutterPluginBinding binding) {}
 
+  final PluginRegistry.ActivityResultListener resultListener =
```
Can you please move this definition up, before it is used for the first time?

kennethj replied:
moved up

mehmetf commented on packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java:
```diff
@@ -65,138 +85,275 @@ private LocalAuthPlugin(Activity activity) {
 
+  private LocalAuthPlugin(Registrar registrar) {
+    this.registrar = registrar;
+    this.registrar.addActivityResultListener(resultListener);
```
There's no need for this constructor.  You can do all this in the registerWith() call.

kennethj replied:
Moved code to `registerWith()` method and deleted extra constructor.

mehmetf commented on packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java:
```diff
@@ -29,14 +39,28 @@
+    if (registrar != null) {
+      return registrar.activity();
+    }
+    return null;
```
Should  you ever be here? I don't see you checking  for null in everywhere getActivity is called  below.

mehmetf commented on packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java:
```diff
@@ -65,138 +85,275 @@ private LocalAuthPlugin(Activity activity) {
+
+  private boolean isDeviceSupported() {
+    KeyguardManager keyguardManager =
+        (KeyguardManager) getActivity().getBaseContext().getSystemService(KEYGUARD_SERVICE);
```
If these services can be grabbed via ApplicationContext, prefer doing that and save  them as instance members to the plugin class. That way you don't have to get them with every call and eliminate the possibility that getActivity could be null.

kennethj replied:
Removed `getActivity()`. Setting services and activity as instances variables when plugin registered and on attached to activity.

mehmetf commented on packages/local_auth/example/lib/main.dart:
```diff
@@ -21,11 +21,21 @@ class MyApp extends StatefulWidget {
-  late bool _canCheckBiometrics;
-  late List<BiometricType> _availableBiometrics;
+
+  bool? _isSupported;
```
Prefer  initializing variables rather than leaving them nullable.

kennethj replied:
I'll add an enum to avoid the `null`

kennethj replied:
Updated to use the `enum` `_SupportState` instead of a nullable `bool`

mehmetf commented on packages/local_auth/lib/auth_strings.dart:
```diff
@@ -15,35 +15,35 @@ import 'package:intl/intl.dart';
-    this.fingerprintHint,
-    this.fingerprintNotRecognized,
-    this.fingerprintSuccess,
+    this.biometricHint,
```
This is a breaking change. Could you please add it to the CHANGELOG?

kennethj replied:
Noted to changelog.

mehmetf commented on packages/local_auth/lib/local_auth.dart:
```diff
@@ -66,7 +66,7 @@ class LocalAuthentication {
   /// [PlatformException] with error code [otherOperatingSystem] on the iOS
   /// simulator.
-  Future<bool> authenticateWithBiometrics({
+  Future<bool?> authenticateWithBiometrics({
```
This should not return a nullable type.

kennethj replied:
Updated to return `Future<bool>` instead of `Future<bool?>`

mehmetf commented on packages/local_auth/lib/local_auth.dart:
```diff
@@ -92,9 +92,77 @@ class LocalAuthentication {
+  /// authentication (e.g. lack of relevant hardware). This might throw
+  /// [PlatformException] with error code [otherOperatingSystem] on the iOS
+  /// simulator.
+  Future<bool?> authenticate({
```
Since you are making a breaking change already due to message strings, why not rename the method `authenticateWithBiometrics` as `authenticate` instead of having a near duplicate method?

That would leave the code much simpler.

kennethj replied:
I wanted to keep the original behavior intact of using biometrics only with `authenticateWithBiometrics` and have `authenticate` use any means available, then perhaps a `authenticateWithCredentials` at some point.

If we remove `authenticateWithBiometrics`, I would want to add a way to keep the original biometric only behavior. Maybe a `bool biometricOnly = false,` parameter on `authenticate`?

mehmetf replied:
Yes, a `biometricOnly` parameter sounds good.

I like your original idea  as well (having separate methods) however the implementation (and API) is largely shared between these methods and you would need to make sure a feature added to `authenticate` is also added to all of its 
 variants (`authenticateWithBiometrics` etc).

You can reduce the implementation of `authenticateWithBiometrics` to just calling `authenticate` with `biometricOnly` set to true. However, I would rather keep the API clean and simple.

kennethj replied:
Added a `biometricOnly` parameter to `authenticate`. `authenticateWithBiometrics` now calls `authenticate` with `biometricOnly: true`

mehmetf reviewed (commented):
Thanks for the contribution (and especially tests). There's additional work required before we can merge this into the repo.

kennethj commented:
@mehmetf, thanks for reviewing that so quickly!

I will test these changes on a few emulators and finish updates based on your review tomorrow.

mehmetf commented:
Thanks for making all these changes. Let me know when it is ready for  the second round!

kennethj commented:
@mehmetf 
I believe I addressed all the comments in the first review and ready for a second look.

Also, I am not sure what's up with the failing CI build `build-apks+java-test+firebase-test-lab CHANNEL:stable PLUGIN_SHARDING:--shardIndex 0 --shardCount 4`. I started updating gradle versions, but that started to become a rabbit hole.

zigapovhe commented:
> @mehmetf
> I believe I addressed all the comments in the first review and ready for a second look.
> 
> Also, I am not sure what's up with the failing CI build `build-apks+java-test+firebase-test-lab CHANNEL:stable PLUGIN_SHARDING:--shardIndex 0 --shardCount 4`. I started updating gradle versions, but that started to become a rabbit hole.

@kennethj Did you enable  'android.useAndroidX' property to true? Set this property to true in the gradle.properties file and retry.

mehmetf commented on packages/local_auth/CHANGELOG.md:
```diff
@@ -28,8 +37,10 @@
 
 ## 0.6.3+1
 
+* Add `authenticate` method that uses biometric authentication, but allows users to also use device authentication - pin, pattern, passcode.
```
We are not going to publish this as all changes are now done in the null-safety version. Please move to the CHANGELOG description above.

kennethj replied:
Moved.

mehmetf commented on packages/local_auth/CHANGELOG.md:
```diff
@@ -61,10 +72,12 @@
 
 * Replace deprecated `getFlutterEngine` call on Android.
 
+
```
Remove these extra newlines. It might be due to formatting done in your editor?

kennethj replied:
extra newlines removed

mehmetf commented on packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java:
```diff
@@ -67,136 +95,255 @@ public LocalAuthPlugin() {}
-        result.error("auth_in_progress", "Authentication in progress", null);
-        return;
-      }
+    this.result = result;
```
This is dangerous. I assume you are saving result because some operation below can respond asynchronously via `onPermissionResult`, `onActivityResult` or some such. You should save the result object on a per-method basis right before you invoke the `startActivityForResult`. Otherwise, another method call can override this result object and you end up calling `result` methods more than once.

This also gives you the chance to reject a second call to the async method. e.g. if you are still awaiting the result of some `foo` method and user calls `foo` again, you can ignore it and instead wait for the `Result  fooResult` to be fulfilled.

mehmetf replied:
I think this is now the only blocker before I LGTM this PR.

kennethj replied:
Removed class level `result` field. Added`lockRequestResult` field that is only set when lock request activity started. This field `lockRequestResult` is only used in activity result delegate function.

mehmetf commented on packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java:
```diff
@@ -67,136 +95,255 @@ public LocalAuthPlugin() {}
+    this.result = result;
+    switch (call.method) {
+      case "authenticate":
+        this.authenticate(call, result);
```
`this.` qualifier is unnecessary (ditto below).

kennethj replied:
Updated

mehmetf commented on packages/local_auth/lib/local_auth.dart:
```diff
@@ -73,28 +73,87 @@ class LocalAuthentication {
     AndroidAuthMessages androidAuthStrings = const AndroidAuthMessages(),
     IOSAuthMessages iOSAuthStrings = const IOSAuthMessages(),
     bool sensitiveTransaction = true,
+  }) =>
```
I would strongly recommend deprecating this method and then removing it sometime in the future.

Please put a `@deprecated` qualifier to it, remove its dartdoc replacing it with "This method is deprecated, please use `authenticate` with `biometricOnly: true`."

kennethj replied:
Added `@Deprecated` annotation and comment docs

mehmetf commented on packages/integration_test/android/build.gradle:
```diff
@@ -8,7 +8,7 @@ buildscript {
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:3.6.3'
+        classpath 'com.android.tools.build:gradle:4.1.1'
```
I am not entirely sure how version resolving for gradle works for this testbed. I will get someone from Flutter core team to review these particular changes. That's likely what's causing your build failure.

mehmetf replied:
@cyanglaz Could you please take a look at the build file changes in this PR? What is the proper way to up the gradle version?

kennethj replied:
I revert the changes to the `integration_test` package in this PR. This appears to prevent those failed builds.

mehmetf commented:
Published as 1.1.0-nullsafety.

stuartmorgan commented on packages/local_auth/ios/Classes/FLTLocalAuthPlugin.m:
```diff
@@ -130,6 +136,48 @@ - (void)authenticateWithBiometrics:(NSDictionary *)arguments
+  NSError *authError = nil;
+  _lastCallArgs = nil;
+  _lastResult = nil;
+  context.localizedFallbackTitle = @"";
```
Why is fallback being disabled here? That seems undesirable for the non-biometric-only case.

mtskf replied:
Please escalate this!
